controlplane/kubeadm/internal/controllers/helpers_test.go

```diff
@@ -539,53 +555,132 @@ func TestKubeadmControlPlaneReconciler_generateMachine(t *testing.T) {
+			infraRef, bootstrapRef,
+			failureDomain, nil,
+		)
+		g.Expect(err).To(BeNil())
+
+		expectedMachineSpec := clusterv1.MachineSpec{
+			ClusterName: cluster.Name,
+			Version:     pointer.String(kcp.Spec.Version),
+			Bootstrap: clusterv1.Bootstrap{
+				ConfigRef: bootstrapRef.DeepCopy(),
+			},
+			InfrastructureRef:       *infraRef.DeepCopy(),
+			FailureDomain:           failureDomain,
+			NodeDrainTimeout:        kcp.Spec.MachineTemplate.NodeDrainTimeout.DeepCopy(),
+			NodeDeletionTimeout:     kcp.Spec.MachineTemplate.NodeDeletionTimeout.DeepCopy(),
+			NodeVolumeDetachTimeout: kcp.Spec.MachineTemplate.NodeVolumeDetachTimeout.DeepCopy(),
+		}
+		g.Expect(createdMachine.Name).To(HavePrefix(kcp.Name))
+		g.Expect(createdMachine.Namespace).To(Equal(kcp.Namespace))
+		g.Expect(createdMachine.OwnerReferences).To(HaveLen(1))
+		g.Expect(createdMachine.OwnerReferences).To(ContainElement(*metav1.NewControllerRef(kcp, controlplanev1.GroupVersion.WithKind("KubeadmControlPlane"))))
+		g.Expect(createdMachine.Spec).To(Equal(expectedMachineSpec))
+
+		// Verify that the machineTemplate.ObjectMeta has been propagated to the Machine.
+		for k, v := range kcpMachineTemplateObjectMeta.Labels {
+			g.Expect(createdMachine.Labels[k]).To(Equal(v))
+		}
+		g.Expect(createdMachine.Labels[clusterv1.ClusterNameLabel]).To(Equal(cluster.Name))
+		g.Expect(createdMachine.Labels[clusterv1.MachineControlPlaneLabel]).To(Equal(""))
+		g.Expect(createdMachine.Labels[clusterv1.MachineControlPlaneNameLabel]).To(Equal(kcp.Name))
+
+		for k, v := range kcpMachineTemplateObjectMeta.Annotations {
+			g.Expect(createdMachine.Annotations[k]).To(Equal(v))
+		}
+
+		// Verify the new Machine has the ClusterConfiguration annotation.
+		g.Expect(createdMachine.Annotations).To(HaveKey(controlplanev1.KubeadmClusterConfigurationAnnotation))
+
+		// Verify that machineTemplate.ObjectMeta in KCP has not been modified.
+		g.Expect(kcp.Spec.MachineTemplate.ObjectMeta.Labels).NotTo(HaveKey(clusterv1.ClusterNameLabel))
+		g.Expect(kcp.Spec.MachineTemplate.ObjectMeta.Labels).NotTo(HaveKey(clusterv1.MachineControlPlaneLabel))
+		g.Expect(kcp.Spec.MachineTemplate.ObjectMeta.Labels).NotTo(HaveKey(clusterv1.MachineControlPlaneNameLabel))
+		g.Expect(kcp.Spec.MachineTemplate.ObjectMeta.Annotations).NotTo(HaveKey(controlplanev1.KubeadmClusterConfigurationAnnotation))
+	})
+
+	t.Run("should return the correct Machine object when updating an existing Machine", func(t *testing.T) {
+		g := NewWithT(t)
+
+		machineName := "existing-machine"
+		machineUID := types.UID("abc-123-existing-machine")
+		clusterConfigurationString := "cluster-configuration-information"
+		failureDomain := pointer.String("fd-1")
+		machineVersion := pointer.String("v1.25.3")
+		exitingMachine := &clusterv1.Machine{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: machineName,
+				UID:  machineUID,
+				Annotations: map[string]string{
+					controlplanev1.KubeadmClusterConfigurationAnnotation: clusterConfigurationString,
```

Comment:
let's also add the remediation annotation  (would have found the bug I found there)